File: src/main/scala/nu/tengstrand/tetrisanalyzer/move/Move.scala
package nu.tengstrand.tetrisanalyzer.move

import rotation.RotationDirection
import nu.tengstrand.tetrisanalyzer.settings.rotation.MoveAdjustment

object Move {
  def apply(rotation: Int, x: Int, y: Int) = new Move(rotation, x, y)
}

/**
 * Holds a position (x,y) and rotation ("angle") for a piece on the board.
 */
class Move(val rotation: Int, val x: Int, val y: Int) {
  def up = new Move(rotation, x, y-1)
  def down = new Move(rotation, x, y + 1)
  def left = new Move(rotation, x - 1, y)
  def right = new Move(rotation, x + 1, y)
  def rotate(adjust: MoveAdjustment, rotationDirection: RotationDirection, modulus: Int) = new Move((rotation + rotationDirection.direction) & modulus, x + adjust.x, y + adjust.y)

  override def equals(that: Any) = that match {
    case other: Move => x == other.x && y == other.y && rotation == other.rotation
    case _ => false
  }

  override def hashCode = (x + y*1000 + rotation * 10000).hashCode
  override def toString = "(" + rotation + "," + x + ", " + y + ")"
}

File: blog/03/src/main/java/tetrisanalyzer/Move.java
package tetrisanalyzer;

public class Move {
    public final int rotation;
    public final int x;
    public final int y;

    public Move(int rotation, int x, int y) {
        this.rotation = rotation;
        this.x = x;
        this.y = y;
    }

    public Move left() {
        return new Move(rotation, x - 1, y);
    }

    public Move right() {
        return new Move(rotation, x + 1, y);
    }

    public Move rotate() {
        return new Move(((rotation+1) & 1), x, y);
    }

    public Move down() {
        return new Move(rotation, x, y + 1);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Move move = (Move) o;

        if (rotation != move.rotation) return false;
        if (x != move.x) return false;
        if (y != move.y) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = rotation;
        result = 31 * result + x;
        result = 31 * result + y;
        return result;
    }

    @Override
    public String toString() {
        return "Move{" +
                "rotation=" + rotation +
                ", x=" + x +
                ", y=" + y +
                '}';
    }
}
